SSA/Core/Util.lean

```diff
@@ -109,3 +109,45 @@ instance [inst : Cli.ParseableType τ] {n : ℕ} : Cli.ParseableType (Vector τ
       return h ▸ Vector.ofArray arr
     else
       none
+
+
+def Std.BitVec.width : Std.BitVec w → Nat := fun _ => w
+
+-- From Haskell:
+def List.toTiers : List α  → List (List α) := (List.groupBy (fun _ _ => false))
+
+def appendTiers : List (List α) → List (List α) → List (List α)
+  | xss, []  =>  xss
+  | [] ,  yss  => yss
+  | (xs::xss), (ys::yss) =>  (xs ++ ys) :: (appendTiers xss yss)
+
+--notation "\\/" => appendTiers
+
+-- fugly; sorry, Lean doesn't have that neat list comprehension
+/-- In Haskell this was defined inline as `**` -/
+def prodWith : (α → β → γ) → List α → List β → List γ
+  | f, as, bs => Id.run do
+      let mut res := []
+      for a in as do
+        for b in bs do
+         res := res ++ [ f a b]
+      res
```

Comment:
Isn't this just `List.zip as bs |>.map f`?

Reply:
@alexkeizer no it is not, note that `zip` will have `n` elements, while this has `n^2` elements.

Reply:
Ahh, now I see: it's taking the cartesian product, rather than zipping the two lists, right

Reply:
```suggestion
def prodWith (f : α → β → γ) (as : List α) (bs : List β) : List γ :=
  as.bind fun a => bs.map fun b => f a b

example : prodWith (String.append) ["a", "b"] ["c", "d"] =
  ["ac", "ad", "bc", "bd"] := by decide
```